api/controller/users.js

```diff
@@ -49,6 +52,218 @@ exports.findById = function(req, res) {
 	});
 };
 
+// function to generate OTP Token for QR code.
+function generateOTPToken (username, serviceName, secret){
+	return otplib.authenticator.keyuri(
+		username,
+		serviceName,
+		secret
+	);
+}
+
+exports.updateSecret = function(req, res){
+	if (!mongo.ObjectID.isValid(req.user._id)) {
+		res.status(401).send({
+			'error': 'Invalid user id',
+			'code': 8
+		});
+		return;
+	}
+
+	var uid = req.user._id;
+
+	// Save unique secret in database.
+	db.collection(usersCollection, function(err, collection) {
+		collection.findOne({
+			_id: new mongo.ObjectID(uid),
+		}, function(err, user) {
+			// only update database with unique secret if tfa is false or not defined -- for existing users in databse.
+			if (user.tfa === false || typeof user.tfa === "undefined") {
+				var uniqueSecret = otplib.authenticator.generateSecret();
+				db.collection(usersCollection, function(err, collection) {
+					collection.findOneAndUpdate({
+						'_id': new mongo.ObjectID(uid)
+					}, {
+						$set:
+						{
+							uniqueSecret: uniqueSecret
+						}
+					}, {
+						safe: true,
+						returnOriginal: false
+					}, function(err, result) {
+						if (err) {
+							res.status(500).send({
+								'error': 'An error has occurred',
+								'code': 7
+							});
+						} else if (result) {
+							var updatedUser = result.value;
+							var name = updatedUser.name;
+							var manualKey = updatedUser.uniqueSecret;
+							var otpAuth = generateOTPToken(name, serviceName, manualKey);
+
+							delete updatedUser.password;
+							//send user, manualKey and otpAuth for QRCode async function.
+							res.send({
+								user: user,
+								uniqueSecret: manualKey,
+								otpAuth: otpAuth
+							});
+						} else {
+							res.status(401).send({
+								'error': 'Inexisting user id',
+								'code': 8
+							});
+						}
+					});
+				});
+			} else if (user.tfa === true) {
+				res.status(500).send({
+					'error': 'An error has occurred',
+					'code': 7
+				});
+			}
+		});
+	});
+
+};
+
+exports.verifyTOTP = function(req, res) {
+	//validate
+	if (!mongo.ObjectID.isValid(req.user._id)) {
+		res.status(401).send({
+			'error': 'Invalid user id',
+			'code': 8
+		});
+		return;
+	}
+
+	var uid = req.user._id;
+
+	if (!req.body.userToken) {
+		res.status(401).send({
+			'error': 'User Token not defined',
+			'code': 31
+		});
+		return;
+	}
+
+	var uniqueToken = req.body.userToken;
+
+	db.collection(usersCollection, function(err, collection) {
+		collection.findOne({
+			_id: new mongo.ObjectID(uid),
+		}, function(err, user) {
+			var uniqueSecret = user.uniqueSecret;
+			if (!err) {
+				try {
+					var isValid = otplib.authenticator.check(uniqueToken, uniqueSecret);
+				} catch (err) {
```

Comment:
Add a return statement to skip checking the following conditions.

Reply:
I didn't get you here, can you please elaborate a bit more?